contracts/test/LONStaking.t.sol

```diff
@@ -433,80 +534,315 @@ contract LONStakingTest is Test {
 
         uint256 lonStakingXLONBefore = lonStaking.totalSupply();
-        BalanceSnapshot.Snapshot memory lonStakingLON = BalanceSnapshot.take(address(lonStaking), address(lon));
-        BalanceSnapshot.Snapshot memory userXLON = BalanceSnapshot.take(user, address(lonStaking));
-        BalanceSnapshot.Snapshot memory userLON = BalanceSnapshot.take(user, address(lon));
 
         uint256 redeemAmount = lonStaking.balanceOf(user);
-        uint256 expectedLONAmount = _getExpectedLONWithoutPenalty(redeemAmount);
-        lonStaking.redeem(redeemAmount);
-        vm.stopPrank();
+        _redeemAndValidate(user, redeemAmount);
 
         uint256 lonStakingXLONAfter = lonStaking.totalSupply();
         assertEq(lonStakingXLONAfter, lonStakingXLONBefore - redeemAmount);
-        lonStakingLON.assertChange(-int256(expectedLONAmount));
-        userXLON.assertChange(-int256(redeemAmount));
-        userLON.assertChange(int256(expectedLONAmount));
+    }
+
+    function testFuzz_RedeemMultipleWithBuyback_OneByOne(uint80[2][16] memory stakeAndRedeemAmounts, uint80[16] memory buybackAmounts) public {
+        // LON cap lies between 2**87 and 2**88, setting upper bound of stake amount to 2**80 - 1 so stake amount will not exceed cap
+        uint256 totalLONAmount = lon.totalSupply();
+        (uint256[16] memory stakeAmounts, uint256[16] memory redeemAmounts) = _determineStakeRedeemAmount(stakeAndRedeemAmounts);
+        for (uint256 i = 0; i < stakeAmounts.length; i++) {
+            uint256 stakeAmount = stakeAmounts[i];
+            uint256 redeemAmount = redeemAmounts[i];
+            vm.assume(stakeAmount > 0);
+            vm.assume(redeemAmount > 0);
+            vm.assume(stakeAmount > redeemAmount);
+            totalLONAmount = totalLONAmount.add(stakeAmount);
+            vm.assume(totalLONAmount <= lon.cap());
+        }
+        for (uint256 i = 0; i < buybackAmounts.length; i++) {
+            vm.assume(buybackAmounts[i] > 0);
+            totalLONAmount = totalLONAmount.add(buybackAmounts[i]);
+            vm.assume(totalLONAmount <= lon.cap());
+        }
+
+        // Make initial big enough deposit so LON amount will not increase dramatically relative to xLON amount due to buyback
+        // and hence result in later staker getting zero shares
+        lon.mint(address(0x5566), 10_000_000e18);
+        _stake(address(0x5566), 10_000_000e18); // stake 10m LON
+
+        for (uint256 i = 0; i < stakeAmounts.length; i++) {
+            address staker = address(uint256(fuzzingUserStartAddress) + i);
+            uint256 stakeAmount = stakeAmounts[i];
+            uint256 redeemAmount = redeemAmounts[i];
+            lon.mint(staker, stakeAmount);
+            _stake(staker, stakeAmount);
+            simulateBuyback(buybackAmounts[i]);
+            // Skip if stake did not get any share due to too small stakeAmount and rounding error
+            if (lonStaking.balanceOf(staker) == 0) continue;
```

Comment:
I realized that if `stakeAmount` is too small and buyback amounts quickly add up, it still could result in staker getting zero share (despite having staked 10m LON first).
For example, stake `1` would still result in getting `1.mul(10000).div(10001) == 0` share

I tried to check `vm.assume(stakeAmount > 10)` but it results in longer test time (fuzzing likely spend much time looking at edge cases).